### PR TITLE
Fix #12442: Confusing error message when the intro pattern of "apply in" fails

### LINKED ISSUE #12442: Confusing error message when the intro pattern of "apply in" fails
#### Description of the problem

When `apply in` is used with a tuple and the intro pattern fails, the error message incorrectly states that the application failed:

```coq
Parameter A B : Prop.
Axiom P : inhabited (A -> B).

Lemma bug (a : A) : True.
  Fail apply P in a as [].
  (* Unable to apply lemma of type "inhabited (A -> B)" on hypothesis of type "A". *)
  apply P in a. (* success *)
```
Note that this bug does not occur when there is no tuple:
```coq
Parameter A B : Prop.
Axiom P' : A -> B.

Lemma ok (a : A) : True.
  Fail apply P' in a as [].
  (* Not an inductive product. *)
```

#### Coq Version
8.11.0

### PR DESCRIPTION
**Pull request by gasparattila**
**Kind:** bug fix

Fixes / closes #12442

- [x] Added / updated test-suite

### CODE REVIEW

File: tactics/tactics.ml [hunk 1]
```diff
@@ -1645,6 +1645,9 @@ let make_projection env sigma params cstr sign elim i n c u =
       | None -> None
   in elim
 
+(* Used for detecting failure of the intro pattern with "apply in" *)
+exception FailureAfterApply of exn
```
Comment (ppedrot) [inline tactics/tactics.ml:1649]: Wrapping exceptions is deemed dangerous in the Coq source code. I wonder whether this shouldn't be done in a more functional way, e.g. returning an algebraic type in the related functions.

**Review by ppedrot — APPROVED** [review]
This is more contrived than I'd have hoped, but I guess there is no way around judging from the spaghetti implementation of the `apply` tactical.

**Comment by ppedrot** [comment]
Just in case I am running [a bench](https://ci.inria.fr/coq/view/benchmarking/job/benchmark-part-of-the-branch/949/console). I'll merge this after leaving a bit of time for people to optionally comment.

**Comment by herbelin** [comment]
The PR is ok to me. 

> Is the code really spaghetti? 

I would rather call it continuation-passing-style, with, for each lemma of the list, finding a component of the tuple, such that unification works, chaining these lemmas on the first subgoal, finally applying an introduction pattern.

@ppedrot, would you prefer to have:
```
tclTHENLAST (tclMAPLAST (apply_in_wo_cps id) lemmas) (intro_patterns pat) id
```
for some new `tclMAPLAST` chaining tactics on the last subgoal  and `apply_in_wo_cps` a non-cps version of `apply_in`, rather than the current (simplified for better exposure)
```
List.fold_right (apply_in_delayed_once id) lemmas (intro_patterns pat) id
```
where `apply_in_delayed_once id lem` of type `unit tactic -> unit tactic`, is chaining tactics.

That would make the flow for error reporting more natural.

**Review by ppedrot — APPROVED** [review]
This looks way simpler!

**Comment by herbelin** [comment]
> This looks way simpler!

You're referring to the last changes in the PR or to my message?

**Comment by ppedrot** [comment]
@herbelin the latest state of the PR.

**Comment by Zimmi48** [comment]
@ppedrot You can still merge PRs in 8.12+beta1 until we close the milestone.

**Pull request merged**